### DISCUSSION TITLE
Rendered more hooks than during the previous render.

### DISCUSSION THREAD

**Discussion by wyf19870504** [vercel/next.js#64157] [Help]
### Summary

The following code in a next.js project, I will get a global-error "Rendered more hooks than during the previous render."  when  ` router.push('/login')` runs.  
If I don't jump the router, or throw an error in "then", it won't make an error, but I want to know why.  T-T
```
//page.tsx
import ClientDemo from './client';
async function queryData() {
  'use server';
}

export default function Page() {
  return (
      <ClientDemo queryData={queryData} />
  );
}
```

```
//client.tsx
'use client';
import { useEffect } from 'react';

export default function ClientDemo(props: any) {
  useEffect(() => {
    throw new TypeError('test');
  });
  props.queryData().then(() => {});
  return (
      <div>...</div>
  );
}
```

```
//error.tsx
'use client';

import { useRouter } from 'next/navigation';
import { useEffect } from 'react';

export default function Error({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
  const router = useRouter();
  useEffect(() => {
    const timeoutId = setTimeout(() => {
      router.push('/login');    //I will get an error "Rendered more hooks than during the previous render." here
    }, 5000);
  }, []);
  return (
    <div>
      <h2>Something was wrong.</h2>
    </div>
  );
}
```

### Additional information

```js
Next version: 14.1.0
```

**Comment by icyJoseph** [comment]
I think the problem has to be with calling the server action as you render, rather than in a callback or use-effect.

**Reply by wyf19870504** [reply]
But If I replace "throw new TypeError('test');" with "router.push" in the useEffect of client.tsx, it won't get error.

**Reply by icyJoseph** [reply]
Hi,

Yeah, because router.push is a side-effect, so you are actually expected to now call it within the render pass, that is, the top level scope of a React Component. Remember that, rendering has to be as pure as possible, without side-effects.
